File: learning_qt_models/custom_list_model.py
from PySide6.QtCore import QAbstractListModel, Qt

class CustomListModel(QAbstractListModel):
    def __init__(self, data=None):
        super().__init__()
        self._data = data or []

    def rowCount(self, parent=None):
        return len(self._data)

    def data(self, index, role=Qt.DisplayRole):
        if not index.isValid():
            return None
        
        if role == Qt.DisplayRole:
            return self._data[index.row()]
        
        if role == Qt.DecorationRole:
            # Ví dụ: trả về màu sắc hoặc icon tùy logic
            pass
            
        return None

    # Phương thức để thêm dữ liệu
    def add_item(self, item):
        self.beginInsertRows(self.index(0).parent(), self.rowCount(), self.rowCount())
        self._data.append(item)
        self.endInsertRows()

File: learning_qt_models_cpp/custom_list_model.cpp
#include "custom_list_model.h"

CustomListModel::CustomListModel(const QStringList &data, QObject *parent)
    : QAbstractListModel(parent), m_data(data)
{
}

int CustomListModel::rowCount(const QModelIndex &parent) const
{
    if (parent.isValid())
        return 0;
    return m_data.count();
}

QVariant CustomListModel::data(const QModelIndex &index, int role) const
{
    if (!index.isValid())
        return QVariant();

    if (index.row() >= m_data.size())
        return QVariant();

    if (role == Qt::DisplayRole)
        return m_data.at(index.row());

    return QVariant();
}

void CustomListModel::addItem(const QString &item)
{
    beginInsertRows(QModelIndex(), rowCount(), rowCount());
    m_data.append(item);
    endInsertRows();
}
